torch/autograd/gradcheck.py

```diff
@@ -334,18 +358,42 @@ def combine_jacobian_rows(jacobians_rows, inputs, numel_outputs) -> Tuple[Tuple[
     return out_jacobians, correct_grad_sizes, correct_grad_types
 
 
+FAILED_NONDET_MSG = """
+
+NOTE: If your op relies on non-deterministic operations i.e., it is listed here:
+https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html
+this failure might be expected.
+
+If you are adding a new operator, please file an issue and then use one of the
+workarounds. The workaround depends on how your test invokes gradcheck/gradgradcheck.
+If the test
+- manually invokes gradcheck/gradgradcheck, then call gradcheck/gradgradcheck
+  with `nondet_tol=<tol>` as a keyword argument.
+- is OpInfo-based (e.g., in test_ops.py), then modify the OpInfo for the test
+  to have `gradcheck_nondet_tol=<tol>`.
+- is a Module test (e.g., in common_nn.py), then modify the corresponding
+  module_test entry to have `gradcheck_nondet_tol=<tol>`
+"""
+
+
 def check_analytical_jacobian_attributes(inputs, output, nondet_tol, grad_out_scale, check_grad_dtypes,
-                                         raise_exception, custom_vjp_fn=None) -> Tuple[Tuple[torch.Tensor, ...], bool]:
+                                         raise_exception, custom_vjp_fn=None, fast_mode=False,
+                                         v=None) -> Tuple[Tuple[torch.Tensor, ...], bool]:
     diff_input_list = list(iter_tensors(inputs, True))
 
     def backward_fn(grad_output):
         return torch.autograd.grad(output, diff_input_list, grad_output,
                                    retain_graph=True, allow_unused=True)
     vjp_fn = custom_vjp_fn if custom_vjp_fn is not None else backward_fn
-    jacobians_rows = compute_analytical_jacobian_rows(vjp_fn, output.clone(), grad_out_scale)
-    jacobians_rows_reentrant = compute_analytical_jacobian_rows(vjp_fn, output.clone(), grad_out_scale)
 
-    output_numel = output.numel()
+    if fast_mode:
+        jacobians_rows = compute_fast_analytical_jacobian_rows(vjp_fn, output.clone(), v, grad_out_scale)
+        jacobians_rows_reentrant = compute_fast_analytical_jacobian_rows(vjp_fn, output.clone(), v, grad_out_scale)
+    else:
+        jacobians_rows = compute_analytical_jacobian_rows(vjp_fn, output.clone(), grad_out_scale)
+        jacobians_rows_reentrant = compute_analytical_jacobian_rows(vjp_fn, output.clone(), grad_out_scale)
```

Comment:
This is a little weird to me:
-  When we're doing slow gradcheck, the quantity that comes out is mapping from `input index -> row of the jacobian -> value`
- When we're doing fast gradcheck, the quantity that comes out is a vjp, which is not a row of the jacobian. It looks like that data structure is a mapping from `input index -> nothing -> vjp`

Further down the line, we're doing `combine_jacobian_rows`, which I have a hard time wrapping my head around -- what does "combine_jacobian_rows" do if `jacobian_rows` isn't actually a mapping from `input index -> row of the jacobian -> value`?

Maybe these just need better naming?
- compute_fast_analytical_jacobian_rows is not actually computing the rows of the jacobian. Maybe we should call it `compute_vjp`
- combine_jacobian_rows -> "stack tensors"?? Not really sure

Reply:
Yeah `compute_vjp` and something like `stack_and_check_tensors` definitely better reflects what the code is doing if we consider both the fast and slow cases. Maybe even `compute_vjp` -> `get_vjp_wrt_specific_output` as to mirror the numerical case.

I'm probably going to still have both if and else update the same variable `jacobians_rows` though. I don't see a better name at the moment and it still makes _some_ sense if you think of the vjp as a linear combination of the jacobian's rows. Adding a comment in the code to explain this.